### ISSUE TITLE
Unable to expand wildcard path

### ISSUE THREAD

**Issue by arielnmz** [davestewart/vuex-pathify#53]
After upgrading a working project from nuxt 1.4 to 2.5 I'm getting this error

> [Vuex Pathify] Unable to expand wildcard path ...:
> - The usual reason for this is that the router was set up before the store
> - Make sure the store is imported before the router, then reload

This is how I added pathify to the vuex plugins:

```
import pathify from 'vuex-pathify'
...
export const plugins = [
  createPersistedState({
    key: 'myapp',
    paths: ['session',]
  }),
  pathify.plugin,
]
```

I'm using nuxt in spa mode

**Comment by davestewart** [comment]
Hmmm. I can't say why that's not working. 

Can you post in the Nuxt issues and see what comes up?

**Comment by arielnmz** [comment]
I'm trying to replicate it with a clean, new nuxt app.

**Comment by arielnmz** [comment]
I've created a new nuxt app and set up a simple store with the following structure:

```

export const state = () => ({
  ui: {
    sub: {
      attr: false,
      ...
    },
  },
})

```
And I'm trying to use it like this in my component

```

  export default {
    name: 'MyComponent',
    computed: {
      ...sync('ui@sub.*')
    }
  }
```

And this is the error I get:

>  Unable to expand wildcard path 'ui@sub.*':

Changing it to `ui@sub` works. It used to work in version `1.1.3` with nuxt `1.4.5`, is this syntax no longer supported?

**Comment by davestewart** [comment]
OK, if you share the nuxt project I'll load it up and take a look.

**Comment by davestewart** [comment]
Is it related to this at all?

https://github.com/davestewart/vuex-pathify/pull/46

Apparently Nuxt 2.x was doing some weird things. I didn't check, just trusted the contributor.

**Comment by arielnmz** [comment]
Finally, I was able to look closer into this. There are tons of scenarios (nuxt in ssr mode, client mode, loading pathify as a client plugin, as aserver plugin, as a vuex plugin). I'll explain a simple and the most common case, assuming we are using `nuxt@^2.5.0` in ssr (universal) mode, and `pathify@^1.2.0`:

If I try to e.g. `sync` a path like `'ui@sub.*'`, pathify tries to expand it via `expandSync` and fails because `vuex.store.state` is `null` at that moment. The reason a path like `ui@sub` doesn't fail is because pathify is not trying to expand it, instead it just calls `getOne`, `syncOne`, etc., which don't need the state to be initialized (they just map the path to the corresponding getter/setter). The problem with the resolver function is that it checks if the state is initialized at the moment of expanding the path, and since it is `null`, it just returns `''` (an empty string), and in the next stage it tries to reduce it, triggering an exception because a string doesn't have the `reduce` method:

> TypeError: paths.reduce is not a function

```
/**
   * Utility function to expand wildcard path for sync()
   *
   * @param   {string}        path        wildcard path
   * @param   {object}        state       state hash
   * @returns {array|string}
   */
  function expandSync (path, state) {
    if (!init(path, state)) {
      return ''
    }
    return resolveStates(path, state)
  }
```

Then:

```
/**
   * Helper function to convert an array of paths to a hash
   *
   * Uses the last path segment as the key
   *
   * @param   {string[]}  paths   An array of paths to convert to a hash
   * @returns {object}            A hash of paths
   */
  function makePathsHash (paths) {
    return paths.reduce(function (paths, path) {
      var key = path.match(/\w+$/);
      paths[key] = path;
      return paths
    }, {})
  }
```

The first thing that comes to mind is to return `[]` instead of `''`, this prevents pathify from crashing but doesn't change the fact that the store wasn't initialized and therefore can't even expand the path.

TL;DR the problem is Nuxt trying to render the routes before initializing the store. Can somebody replicate a working scenario? In the meantime, the "solution" would be to avoid using wildcards.

**Comment by arielnmz** [comment]
Going a step further, this seems to happen only when pathify is used in a layout or any component in a layout

https://codesandbox.io/s/ojqkwonl7y

**Comment by arielnmz** [comment]
Bug report in nuxt's github https://github.com/nuxt/nuxt.js/issues/5405

**Comment by davestewart** [comment]
Hey, thanks for investigating!

I know it's not fun when the codebase is not your own.

Out of interest (and I assume you used the source) how did you find navigating around the source?

Did it seem logical?

**Comment by arielnmz** [comment]
I actually just put some breakpoints in the bundled script to find what's it doing, but I'm seeing the source now, and yes, I can say it's pretty well organized, I quickly found my way around. Ty for your time

**Comment by davestewart** [comment]
FYI, I found a really cool Chrome extension the other day called OctoLinker which adds live links to GitHub repos:

- https://octolinker.github.io/

It's great for this kind of thing!

**Comment by arielnmz** [comment]
Looks great! Im going to give it a try, ty

**Comment by chiboreache** [comment]
> 2.6.1

wildcard error still here

**Comment by davestewart** [comment]
Nuxt 2.6.1 ?

**Comment by chiboreache** [comment]
Yep

![Screenshot_20190408_172807](https://user-images.githubusercontent.com/16459251/55731917-af205e00-5a23-11e9-9401-8d3bd697d230.png)

**Comment by davestewart** [comment]
Ah I see. 

So you mean with the VP 1.2.2 update, Nuxt still has its errors.

You don't mean that Nuxt have released something, but it didn't fix it.

**Comment by chiboreache** [comment]
> So you mean with the VP 1.2.2 update, Nuxt still has its errors.

Right  : ) 

![Screenshot_20190408_180750](https://user-images.githubusercontent.com/16459251/55735525-278a1d80-5a2a-11e9-8dee-f1324510a955.png)


Even after reverting back to 2 version

> Since core-js@2 and core-js@3 are both supported from babel 7.4.0, Starting from 2.6 Nuxt supports both versions (#5411). It is highly advised to stick with 2 (default) unless having a special dependency that needs 3. (Instructions for core-js@2 are not required but recommended)
https://github.com/nuxt/nuxt.js/releases

it doesn't work `~_~`

**Comment by yowzadave** [comment]
I see this issue with just a vanilla `vue-cli` app when I attempt to use wildcard expansion—no Nuxt, not even vue-router. See this sample repo:

https://github.com/yowzadave/vuex-pathify-test

in the "HelloWorld.vue" component (https://github.com/yowzadave/vuex-pathify-test/blob/master/src/components/HelloWorld.vue), the wildcard `get` fails with the above error. The non-wildcard `get` (commented out here) works fine.

**Comment by VesterDe** [comment]
I believe I'm having a related issue.

The error happens if I import my router into a module file in any way. I suppose it's because the store theoretically isn't initialised in this file, it's just a file that exports one of the modules of the state. I don't (yet) do anything with the router... Just importing it causes the error.

Is there any way to get it to work?

My current workaround is to do the things in components that I would rather do in actions.

Thanks.

**Comment by davestewart** [comment]
Hi @VesterDe, 

This is probably an order-of-import issue.

As soon as you import your router file, webpack will then try to import everything that is included in that file, so you end up with a race condition because suddenly your components are being imported before the store they depend on.

You'll have to think of a way to allow the store to finish its setup before using the router, or at least the routes and the dependent components it depends on. Consider adding routes dynamically, lazy-loading of routes or something like this:

```
// helper
function router () {
  return require('/router').default
}

// use router in an action
const actions {
  doSomething () {
    router().whatever()
  }
}
```

Let me know how you get on :)

**Comment by VesterDe** [comment]
Thanks for the feedback, it seemed like what you describe to me too.

I've now solved it the way you recommended, although I'm not crazy about how it looks...

I do prefer to handle the redirect inside actions though, so I'll keep it for now.

For anyone wondering how exactly it looks for me, it's:

```
const router = () => import("@plugins/router");

// Inside my login action.
router().then( ( routerModule ) => {
    if(routerModule.default.app._route.query.loginRedirect){
        routerModule.default.push(routerModule.default.app._route.query.loginRedirect);
    }
});
```

**Comment by davestewart** [comment]
Another option would be moving all that code to a helper in the router folder, then you only need do:

```js
import { isLoggingIn } from '@plugins/router/helpers'

const actions = {
  someAction () {
    if(!isLoggingIn()) {
      // your code
    }
  }
}
```

Because `isLoggingIn()` (or whatever you want to call it) is itself a function, it won't be called until you use it.
